src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,391 @@
+            //     + - - - - - - - - - - - - - - - +-------------------------------+
+            //     |                               |Masking-key, if MASK set to 1  |
+            //     +-------------------------------+-------------------------------+
+            //     | Masking-key (continued)       |          Payload Data         |
+            //     +-------------------------------- - - - - - - - - - - - - - - - +
+            //     :                     Payload Data continued ...                :
+            //     + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
+            //     |                     Payload Data continued ...                |
+            //     +---------------------------------------------------------------+
+
+            // Read at least 2 bytes
+            var buffer = await _inbound.ReadAtLeastAsync(2, cancellationToken);
+
+            // Read the opcode
+            var opcodeByte = buffer.ReadBigEndian<byte>();
+            buffer = buffer.Slice(1);
+
+            var fin = (opcodeByte & 0x01) != 0;
+            var opcode = (WebSocketOpcode)((opcodeByte & 0xF0) >> 4);
+
+            // Read the first byte of the payload length
+            var lenByte = buffer.ReadBigEndian<byte>();
+            buffer = buffer.Slice(1);
+
+            var masked = (lenByte & 0x01) != 0;
+            var payloadLen = (lenByte & 0xFE) >> 1;
+
+            // Mark what we've got so far as consumed
+            _inbound.Advance(buffer.Start);
+
+            // Calculate the rest of the header length
+            var headerLength = masked ? 4 : 0;
+            if (payloadLen == 126)
+            {
+                headerLength += 2;
+            }
+            else if (payloadLen == 127)
+            {
+                headerLength += 4;
+            }
+
+            var maskingKey = default(ReadableBuffer);
+            if (headerLength > 0)
+            {
+                buffer = await _inbound.ReadAtLeastAsync(headerLength, cancellationToken);
+
+                // Read extended payload length (if any)
+                if (payloadLen == 126)
+                {
+                    payloadLen = buffer.ReadBigEndian<ushort>();
+                    buffer = buffer.Slice(sizeof(ushort));
+                }
+                else if (payloadLen == 127)
+                {
+                    var longLen = buffer.ReadBigEndian<ulong>();
+                    buffer = buffer.Slice(sizeof(ulong));
+                    if (longLen > int.MaxValue)
+                    {
+                        throw new WebSocketException($"Frame is too large. Maximum frame size is {int.MaxValue} bytes");
+                    }
+                    payloadLen = (int)longLen;
+                }
+
+                // Read masking key
+                if (masked)
+                {
+                    var maskingKeyStart = buffer.Start;
+                    maskingKey = buffer.Slice(0, 4);
```

Comment:
if you snapshot a slice, then do advance/etc, you have **relinquished** that data; unless you are using `Preserve` / `Dispose`, this data is no longer usable. Of course, if you were capturing it into a `uint` it wouldn't be a problem.